File: go/refs.go
// Copyright (c) 2021-2026 Richard Rodger, MIT License

package tabnastoml

import (
	jsonic "github.com/tabnas/jsonic/go"
)

// Object nodes in this port are insertion-ordered *jsonic.OrderedMap values,
// matching the engine's default object node (and the TS port's plain-object
// insertion order). newMap allocates a fresh one; asMap type-asserts a node
// to *OrderedMap. Container nodes flowing in from the shared engine (the map
// and pair rules) are already OrderedMaps, so building ours the same way lets
// @table-bc / the dive+array handlers merge and index them uniformly.
func newMap() *jsonic.OrderedMap { return jsonic.NewOrderedMap() }

func asMap(v any) (*jsonic.OrderedMap, bool) {
	om, ok := v.(*jsonic.OrderedMap)
	return om, ok
}

// makeRefs builds the function reference map that the grammar file
// references via @-prefixed strings. State-action names
// (@<rule>-<bo|ao|bc|ac>) are auto-wired by Jsonic's Grammar() via
// wireStateActions.
func makeRefs() map[jsonic.FuncRef]any {
	return map[jsonic.FuncRef]any{

		// --- Value-match callbacks (datetime / time) ---

		"@isodate-val":   func(res []string) any { return isodateVal(res) },
		"@localtime-val": func(res []string) any { return localtimeVal(res) },

		// --- State actions (auto-wired by rule name convention) ---

		"@toml-bo": jsonic.StateAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			r.Node = newMap()
		}),

		"@table-bo": jsonic.StateAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			r.Node = r.Parent.Node
		}),

		"@table-bc": jsonic.StateAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			if r.U["top_dive"] != nil {
				return
			}
			if r.Child == nil || r.Child == jsonic.NoRule {
				return
			}
			child, okc := asMap(r.Child.Node)
			node, okn := asMap(r.Node)
			if !okc || !okn {
				return
			}
			for _, k := range child.Keys {
				node.Set(k, child.Vals[k])
			}
		}),

		"@table-ac": jsonic.StateAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			// Reset the dive/array counters on the rule that the parser
			// transitions to after this table closes. Mirrors the TS
			// handler that receives `next` as its third arg.
			next := r.Next
			if next != nil && next != jsonic.NoRule {
				n := next.EnsureN()
				n["table_dive"] = 0
				n["table_array"] = 0
			}
		}),

		"@dive-bc": jsonic.StateAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			if r.U["dive_end"] == nil {
				return
			}
			if r.O0 == nil || r.O0 == jsonic.NoToken {
				return
			}
			key, ok := r.O0.Val.(string)
			if !ok {
				return
			}
			if node, ok := asMap(r.Node); ok {
				node.Set(key, r.Child.Node)
			}
		}),

		// --- Alt actions ---

		"@table-dive-start": jsonic.AltAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			r.EnsureU()
			key := tokenString(r.O0)
			parent, ok := asMap(r.Parent.Node)
			if !ok {
				return
			}
			existing, _ := parent.Get(key)

			if r.N["table_array"] > 0 {
				if arr, ok := existing.([]any); ok {
					if len(arr) > 0 {
						if last, ok := asMap(arr[len(arr)-1]); ok {
							r.Node = last
							return
						}
					}
					newM := newMap()
					arr = append(arr, newM)
					parent.Set(key, arr)
					r.Node = newM
					return
				}
			}

			// Plain-table dive into an existing array: track it so later
			// handlers can treat it like [[…]]. Mirrors the TS behavior
			// `r.parent.node[key] || {}` where truthy arrays pass through
			// unchanged.
			if arr, ok := existing.([]any); ok {
				r.Node = arr
				r.U["arr_parent"] = parent
				r.U["arr_key"] = key
				return
			}
			if m, ok := asMap(existing); ok {
				r.Node = m
				return
			}
			newM := newMap()
			parent.Set(key, newM)
			r.Node = newM
		}),

		"@table-dive-mid": jsonic.AltAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			r.EnsureU()
			key := tokenString(r.O0)
			if _, ok := r.Prev.Node.([]any); ok {
				// Extract array from its actual home in the parent map so
				// appends here are visible to later reads. Go slice headers
				// don't share through map values.
				arrParent, _ := asMap(r.Prev.U["arr_parent"])
				arrKey, _ := r.Prev.U["arr_key"].(string)
				var arr []any
				if arrParent != nil {
					av, _ := arrParent.Get(arrKey)
					arr, _ = av.([]any)
				}
				var last *jsonic.OrderedMap
				if len(arr) > 0 {
					last, _ = asMap(arr[len(arr)-1])
				}
				if last == nil {
					last = newMap()
					arr = append(arr, last)
					if arrParent != nil {
						arrParent.Set(arrKey, arr)
					}
					r.Prev.Node = arr
				}
				// An intervening [[a.b]] leaves last[key] as a []any; keep
				// it so the next table-cs-push can append. Mirrors the TS
				// dive-mid's `r.node = r.prev.node[key] || {}` where a
				// truthy array falls straight through.
				lastVal, _ := last.Get(key)
				if nextArr, ok := lastVal.([]any); ok {
					r.Node = nextArr
					r.U["arr_parent"] = last
					r.U["arr_key"] = key
					return
				}
				next, ok := asMap(lastVal)
				if !ok {
					next = newMap()
					last.Set(key, next)
				}
				r.Node = next
				return
			}
			prev, ok := asMap(r.Prev.Node)
			if !ok {
				return
			}
			// Same array-preservation rule when the previous node was a map
			// rather than a slice (e.g. second [[a.b]] after dive-start
			// returns the first a[0] map whose b already holds an array).
			prevVal, _ := prev.Get(key)
			if arr, ok := prevVal.([]any); ok {
				r.Node = arr
				r.U["arr_parent"] = prev
				r.U["arr_key"] = key
				return
			}
			next, ok := asMap(prevVal)
			if !ok {
				next = newMap()
				prev.Set(key, next)
			}
			r.Node = next
		}),

		"@table-key-cs-head": jsonic.AltAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			r.EnsureU()
			key := tokenString(r.O0)
			parent, ok := asMap(r.Parent.Node)
			if !ok {
				return
			}
			existing, _ := parent.Get(key)
			if existing == nil {
				if r.N["table_array"] > 0 {
					arr := []any{}
					parent.Set(key, arr)
					r.Node = arr
					r.U["arr_parent"] = parent
					r.U["arr_key"] = key
				} else {
					m := newMap()
					parent.Set(key, m)
					r.Node = m
				}
				return
			}
			if arr, ok := existing.([]any); ok {
				r.Node = arr
				r.U["arr_parent"] = parent
				r.U["arr_key"] = key
				return
			}
			r.Node = existing
		}),

		"@table-key-cs-tail": jsonic.AltAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			r.EnsureU()
			key := tokenString(r.O0)
			if _, ok := r.Prev.Node.([]any); ok {
				arrParent, _ := asMap(r.Prev.U["arr_parent"])
				arrKey, _ := r.Prev.U["arr_key"].(string)
				var arr []any
				if arrParent != nil {
					av, _ := arrParent.Get(arrKey)
					arr, _ = av.([]any)
				}
				var last *jsonic.OrderedMap
				if len(arr) > 0 {
					last, _ = asMap(arr[len(arr)-1])
				}
				if last == nil {
					last = newMap()
					arr = append(arr, last)
					if arrParent != nil {
						arrParent.Set(arrKey, arr)
					}
					r.Prev.Node = arr
				}
				lastVal, _ := last.Get(key)
				next, ok := asMap(lastVal)
				if !ok {
					next = newMap()
					last.Set(key, next)
				}
				r.Node = next
				return
			}
			prev, ok := asMap(r.Prev.Node)
			if !ok {
				return
			}
			existing, _ := prev.Get(key)
			if existing == nil {
				if r.N["table_array"] > 0 {
					arr := []any{}
					prev.Set(key, arr)
					r.Node = arr
					r.U["arr_parent"] = prev
					r.U["arr_key"] = key
				} else {
					m := newMap()
					prev.Set(key, m)
					r.Node = m
				}
				return
			}
			if arr, ok := existing.([]any); ok {
				r.Node = arr
				r.U["arr_parent"] = prev
				r.U["arr_key"] = key
				return
			}
			r.Node = existing
		}),

		"@table-cs-push": jsonic.AltAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			newM := newMap()
			if arr, ok := r.Prev.Node.([]any); ok {
				arr = append(arr, newM)
				r.Prev.Node = arr
				// The array also lives in its parent map; writing back
				// there keeps both views consistent after slice growth.
				if arrParent, ok := asMap(r.Prev.U["arr_parent"]); ok {
					if arrKey, ok := r.Prev.U["arr_key"].(string); ok {
						arrParent.Set(arrKey, arr)
					}
				}
			}
			r.Node = newM
		}),

		"@pair-key-set": jsonic.AltAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			if r.O0 != nil && r.O0 != jsonic.NoToken {
				r.EnsureU()["key"] = r.O0.Val
			}
		}),

		"@dive-key-dot": jsonic.AltAction(func(r *jsonic.Rule, _ *jsonic.Context) {
			key := tokenString(r.O0)
			parent, ok := asMap(r.Parent.Node)
			if !ok {
				return
			}
			existingVal, _ := parent.Get(key)
			existing, ok := asMap(existingVal)
			if !ok {
				existing = newMap()
				parent.Set(key, existing)
			}
			r.Node = existing
		}),

		// --- Conditions ---

		"@table-top-dive-cond": jsonic.AltCond(func(r *jsonic.Rule, _ *jsonic.Context) bool {
			return r.D == 1 && (r.Prev == nil || r.Prev.Name != "table")
		}),

		"@lte-table-dive": jsonic.AltCond(func(r *jsonic.Rule, _ *jsonic.Context) bool {
			return r.Lte("table_dive", 0)
		}),

		"@lte-table-array-1": jsonic.AltCond(func(r *jsonic.Rule, _ *jsonic.Context) bool {
			return r.Lte("table_array", 1)
		}),

		"@lte-dive-key-1": jsonic.AltCond(func(r *jsonic.Rule, _ *jsonic.Context) bool {
			return r.Lte("dive_key", 1)
		}),

		"@lte-pk": jsonic.AltCond(func(r *jsonic.Rule, _ *jsonic.Context) bool {
			return r.Lte("pk", 0)
		}),

		"@map-is-table-parent": jsonic.AltCond(func(r *jsonic.Rule, _ *jsonic.Context) bool {
			return r.Parent != nil && r.Parent.Name == "table"
		}),

		// --- Dynamic push/replace targets ---

		"@table-end-p": func(r *jsonic.Rule, _ *jsonic.Context) string {
			if r.N["table_array"] > 0 {
				return ""
			}
			return "map"
		},

		"@table-end-r": func(r *jsonic.Rule, _ *jsonic.Context) string {
			if r.N["table_array"] > 0 {
				return "table"
			}
			return ""
		},
	}
}

// tokenString returns a token's value as a string.
func tokenString(t *jsonic.Token) string {
	if t == nil || t == jsonic.NoToken {
		return ""
	}
	if s, ok := t.Val.(string); ok {
		return s
	}
	if t.Src != "" {
		return t.Src
	}
	return ""
}

File: rs/src/refs.rs
// Copyright (c) 2021-2026 Richard Rodger, MIT License

//! The function references the declarative grammar names with `@`.
//!
//! Registered on the instance BEFORE the grammar document is installed,
//! because the document is what looks them up. The state actions are wired
//! by the `@<rule>-<phase>` convention; the alternate actions, conditions
//! and conditional `p:` / `r:` targets are named explicitly.

use std::cell::RefCell;
use std::rc::Rc;

use tabnas::{ActionError, Context, Rule, Tabnas, Value};

use crate::node::{
    self, array_at, last_of_list, merge_into, new_map, path_of, push, read, set_path,
    snapshot_path, table_at, Seg, DEFINE, DESCEND,
};
use crate::strmatcher::make_toml_string_matcher;
use crate::values::{isodate_val, localtime_val};

/// A matched token's value as a table key. A `#ID` or `#ST` key arrives as
/// a string; anything else falls back to the source text, as the Go
/// `tokenString` does.
fn key_of(rule: &mut Rule, context: &mut Context) -> String {
    match rule.resolve_open_value(0, context) {
        Value::String(text) => text,
        Value::Text(text) => text.string,
        _ => rule
            .o0()
            .map(|token| token.src.to_string())
            .unwrap_or_default(),
    }
}

/// The cell a rule's table path is relative to.
fn cell_of(rule: &Rule) -> Rc<RefCell<Value>> {
    Rc::clone(&rule.node)
}

/// The path this rule's PARENT sits at.
fn parent_path(rule: &Rule) -> Vec<Seg> {
    snapshot_path(rule.parent_rule.as_ref())
}

/// The path the rule this one REPLACED sat at.
fn prev_path(rule: &Rule) -> Vec<Seg> {
    snapshot_path(rule.prev_rule.as_ref())
}

fn truthy(value: Option<&Value>) -> bool {
    match value {
        None | Some(Value::Undefined | Value::Null | Value::Bool(false)) => false,
        Some(Value::Number(number)) => 0.0 != *number,
        Some(Value::String(text)) => !text.is_empty(),
        Some(_) => true,
    }
}

/// Install every `@`-named reference the grammar uses.
pub(crate) fn register(parser: &mut Tabnas) {
    register_matchers(parser);
    register_state_actions(parser);
    register_alt_actions(parser);
    register_conditions(parser);
    register_routes(parser);
}

fn register_matchers(parser: &mut Tabnas) {
    // The custom string matcher, through the grammar's
    // `options.lex.match.string.make`.
    parser.lex_match_factory_ref("@make-toml-string-matcher", make_toml_string_matcher);

    // The grammar declares `val: '@isodate-val'` on its regexp date
    // matchers, so the references have to resolve when the document is
    // installed. The context-aware matchers in `datematcher` are what run,
    // exactly as in the other two ports, and these stay the single
    // description of what a match MEANS.
    parser.value_transform_ref("@isodate-val", isodate_val);
    parser.value_transform_ref("@localtime-val", localtime_val);
}

fn register_state_actions(parser: &mut Tabnas) {
    // The document root.
    parser.state_action_ref("@toml-bo", |rule, _context| {
        rule.node = Rc::new(RefCell::new(new_map()));
        set_path(rule, &[]);
        Ok(())
    });

    // Allocate this map's node up front. The JSON core allocates a braced
    // map's node in its `@object$` alternate action, but TOML prepends its
    // own map-open alternates that match first and carry no `@object$`, so
    // without this a pushed `pair` would inherit no node at all.
    //
    // SUFFIXED because the engine resolves a bare `@map-bo` to its OWN
    // builtin first and a registration under that name is shadowed
    // silently.
    parser.state_action_ref("@map-bo/append", |rule, _context| {
        rule.node = Rc::new(RefCell::new(new_map()));
        set_path(rule, &[]);
        Ok(())
    });

    // A table's node is its parent's, which for every table is the
    // document root; the alternate actions then move it to the table being
    // defined. Assigning the path is the whole of it here: the cell is
    // already the parent's, because a pushed or replaced rule shares it.
    parser.state_action_ref("@table-bo", |rule, _context| {
        if let Some(parent) = rule.parent_rule.clone() {
            rule.node = Rc::clone(&parent.node);
        }
        let path = parent_path(rule);
        set_path(rule, &path);
        Ok(())
    });

    // A dive has no lifecycle action in the canonical grammar, because
    // there the engine's own node inheritance says what its node is: the
    // parent's when the rule was pushed, the previous rule's when it was
    // replaced. A path has to be inherited explicitly to say the same
    // thing.
    parser.state_action_ref("@dive-bo", |rule, _context| {
        let path = if rule.prev_rule.is_some() {
            prev_path(rule)
        } else {
            parent_path(rule)
        };
        set_path(rule, &path);
        Ok(())
    });

    // Fold the table's body into the table, and reset the header counters
    // for whatever comes next.
    parser.state_action_ref("@table-bc", |rule, _context| {
        if !truthy(rule.u.get("top_dive")) {
            if let Some(child) = rule.child_rule.clone() {
                let source = child.node.borrow().clone();
                let path = path_of(rule);
                merge_into(&cell_of(rule), &path, &source);
            }
        }

        // The canonical `@table-ac` writes `next.n.table_dive = 0` on the
        // rule the parser moves to. Here `next` arrives as an immutable
        // snapshot, so the reset is made on THIS rule instead, in the
        // before-close phase: the engine builds the replacement rule's
        // counters by sharing this rule's, and it does so after the
        // before-close actions have run. Same rules reset, same values,
        // one phase earlier. No table close alternate reads either
        // counter, so nothing between the two phases can see the
        // difference.
        let counters = rule.n_mut();
        counters.insert("table_dive".to_string(), 0);
        counters.insert("table_array".to_string(), 0);
        Ok(())
    });

    parser.state_action_ref("@dive-bc", |rule, context| {
        if !truthy(rule.u.get("dive_end")) {
            return Ok(());
        }
        let Some(child) = rule.child_rule.clone() else {
            return Ok(());
        };
        let value = child.node.borrow().clone();
        let key = key_of(rule, context);
        let mut path = path_of(rule);
        path.push(Seg::Key(key));
        node::write(&cell_of(rule), &path, value);
        Ok(())
    });
}

fn register_alt_actions(parser: &mut Tabnas) {
    parser.action_with_context("@table-dive-start", |rule, context| {
        let key = key_of(rule, context);
        let cell = cell_of(rule);
        let parent = parent_path(rule);

        let mut candidate = parent.clone();
        candidate.push(Seg::Key(key.clone()));

        let path = if 0 < rule.n.get("table_array").copied().unwrap_or(0)
            && node::is_list(&cell, &candidate)
        {
            last_of_list(&cell, &candidate)
        } else {
            // A plain-table dive into an existing array of tables walks
            // THROUGH it, which is how `[[x]]` followed by `[x.y]` works.
            table_at(&cell, &parent, &key, DESCEND)?
        };
        set_path(rule, &path);
        Ok(())
    });

    parser.action_with_context("@table-dive-mid", |rule, context| {
        let key = key_of(rule, context);
        let cell = cell_of(rule);
        let previous = prev_path(rule);

        let base = if node::is_list(&cell, &previous) {
            last_of_list(&cell, &previous)
        } else {
            previous
        };
        let path = table_at(&cell, &base, &key, DESCEND)?;
        set_path(rule, &path);
        Ok(())
    });

    parser.action_with_context("@table-key-cs-head", |rule, context| {
        let key = key_of(rule, context);
        let cell = cell_of(rule);
        let parent = parent_path(rule);
        let path = if 0 < rule.n.get("table_array").copied().unwrap_or(0) {
            array_at(&cell, &parent, &key)?
        } else {
            table_at(&cell, &parent, &key, DEFINE)?
        };
        set_path(rule, &path);
        Ok(())
    });

    parser.action_with_context("@table-key-cs-tail", |rule, context| {
        let key = key_of(rule, context);
        let cell = cell_of(rule);
        let previous = prev_path(rule);

        let path = if node::is_list(&cell, &previous) {
            let last = last_of_list(&cell, &previous);
            table_at(&cell, &last, &key, DEFINE)?
        } else if 0 < rule.n.get("table_array").copied().unwrap_or(0) {
            array_at(&cell, &previous, &key)?
        } else {
            table_at(&cell, &previous, &key, DEFINE)?
        };
        set_path(rule, &path);
        Ok(())
    });

    parser.action_with_context("@table-cs-push", |rule, context| {
        let cell = cell_of(rule);
        let previous = prev_path(rule);
        if !node::is_list(&cell, &previous) {
            // `[[a]]` where `a` is already a scalar. The array itself is
            // produced by `@table-key-cs-head`, so reaching a non-array
            // here means the key conflicts.
            let key = key_of(rule, context);
            let existing = read(&cell, &previous).unwrap_or(Value::Undefined);
            return Err(ActionError::new(
                "toml_key_conflict",
                format!(
                    "cannot define {key}, it already has the value {}",
                    existing.to_json()
                ),
            ));
        }
        let Some(index) = push(&cell, &previous, new_map()) else {
            return Ok(());
        };
        let mut path = previous;
        path.push(Seg::Index(index));
        set_path(rule, &path);
        Ok(())
    });

    parser.action_with_context("@pair-key-set", |rule, context| {
        let key = rule.resolve_open_value(0, context);
        rule.u_mut().insert("key".to_string(), key);
        Ok(())
    });

    parser.action_with_context("@dive-key-dot", |rule, context| {
        let key = key_of(rule, context);
        let cell = cell_of(rule);
        let parent = parent_path(rule);
        let path = table_at(&cell, &parent, &key, DESCEND)?;
        set_path(rule, &path);
        Ok(())
    });
}

fn register_conditions(parser: &mut Tabnas) {
    parser.alt_condition("@table-top-dive-cond", |rule, _context| {
        let previous = rule
            .prev_rule
            .as_ref()
            .map_or("", |previous| previous.name.as_ref());
        1 == rule.d && "table" != previous
    });
    parser.alt_condition("@lte-table-dive", |rule, _context| {
        rule.lte("table_dive", 0)
    });
    parser.alt_condition("@lte-table-array-1", |rule, _context| {
        rule.lte("table_array", 1)
    });
    parser.alt_condition("@lte-dive-key-1", |rule, _context| rule.lte("dive_key", 1));
    parser.alt_condition("@lte-pk", |rule, _context| rule.lte("pk", 0));
    parser.alt_condition("@map-is-table-parent", |rule, _context| {
        rule.parent_rule
            .as_ref()
            .is_some_and(|parent| "table" == parent.name.as_ref())
    });
}

fn register_routes(parser: &mut Tabnas) {
    // `!r.n.table_array && 'map'` and `r.n.table_array && 'table'`: a
    // falsy answer means "no route", which is `None` here.
    parser.alt_push("@table-end-p", |rule, _context| {
        (0 == rule.n.get("table_array").copied().unwrap_or(0)).then(|| "map".to_string())
    });
    parser.alt_replace("@table-end-r", |rule, _context| {
        (0 < rule.n.get("table_array").copied().unwrap_or(0)).then(|| "table".to_string())
    });
}
